### ISSUE TITLE
BIP329 Labeling

### ISSUE THREAD

**Issue by i5hi** [SatoshiPortal/bullbitcoin-mobile#40]
This update will complete change the way we are handling labeling. 

Here are some notes from our discussions to help us start with implementation:

New label model:

```dart
  enum BIP329Type { tx, address, pubkey, input, output, xpub }

  const factory Bip329Label({
    required BIP329Type type,
    required String ref,
    String? label,
    String? origin,
    bool? spendable,
  })
```
label can be a comma separated list of labels allowing one ref to have many labels. Refs can also share labels.

Wallets now have two additional fields:

```dart
Wallet...
    List<String>? labelTags,
    List<Bip329Label>? bip329Labels,
```

labelTags contains all the existing labels used in a wallet. When a new label is being added to a particular bip329 ref, we will show the existing labelTags as dropdown to easily add an existing label. This will work similar to github issues labels or tags in most platforms like soundcloud. 

We will need a helper function to find all related refs and they will by default inherit the same label.

**Comment by i5hi** [comment]
[BIP329 Reference](https://github.com/bitcoin/bips/blob/master/bip-0329.mediawiki)

**Comment by BullishNode** [comment]
Doest BIP329 labelling take into account whether an address is external or internal?

**Comment by i5hi** [comment]
Nope. Just whether its spendable or not. We will have to keep track separately of whether the address is change.

**Comment by BullishNode** [comment]
We could add ischange label to BIP329 
https://github.com/BullishNode/bips/compare/master...ischange

**Comment by BullishNode** [comment]
FYI @i5hi we can make sure the "frozen" label we have in the wallet is exported as "spendable": "false"

**Comment by i5hi** [comment]
> We could add ischange label to BIP329 [BullishNode/bips@master...ischange](https://github.com/BullishNode/bips/compare/master...ischange)

We could stay by the existing spec if we add the unhardened path to the origin field as well

Something like this:
```
{ "type": "addr", "ref": "bc1q34aq5drpuwy3wgl9lhup9892qp6svr8ldzyy7c", "label": "Address", "origin": "wpkh([d34db33f/84'/0'/0']/0/1)]" }
 ```
This also gives the index and helps to know where to start a wallets address generation from (although in our case bdk does that).

**Comment by i5hi** [comment]
> FYI @i5hi we can make sure the "frozen" label we have in the wallet is exported as "spendable": "false"

`!spendable without origin` = not your address
`!spendable with origin` = your address  and you've wanted it not to be spent (marked as frozen)

?

**Comment by i5hi** [comment]
> > We could add ischange label to BIP329 [BullishNode/bips@master...ischange](https://github.com/BullishNode/bips/compare/master...ischange)
> 
> We could stay by the existing spec if we add the unhardened path to the origin field as well
> 
> Something like this:
> 
> ```
> { "type": "addr", "ref": "bc1q34aq5drpuwy3wgl9lhup9892qp6svr8ldzyy7c", "label": "Address", "origin": "wpkh([d34db33f/84'/0'/0']/0/1)]" }
> ```
> 
> This also gives the index and helps to know where to start a wallets address generation from (although in our case bdk does that).

Although BIP329 does say this:

"If present, the optional origin property must contain an abbreviated output descriptor (as defined by BIP380[[3](https://github.com/BullishNode/bips/blob/master/bip-0329.mediawiki#cite_note-3)]) describing a BIP32 compatible originating wallet, including all key origin information but excluding any actual keys, any child path elements, or a checksum. This property should be used to disambiguate transaction labels from different wallets contained in the same export, particularly when exporting multiple accounts derived from the same seed."

**Comment by i5hi** [comment]
BIP380 does define origin as just fingerprint and hardened path.

so we would need an isChange if we wanted to to know but we can always sync our wallet and know which of those addresses are change.

**Comment by i5hi** [comment]
@mocodesmo 
- [ ] bip329 methods
implement read/writeToFile for List<Bip329> 
File naming convention,for now, can be `${wallet.id}.bip329`

implement encrypt/decrypt for List<Bip329> (for now use a fixed key of `bullbitcoin-test`) I will later update this to use a key derived from the seed linked to the wallet. 

- [ ] wallet methods
`importBIP329` which will read through wallet.transactions,myAddressBook,externalAddressBook and update only the `label` field based on the List<Bip329>. Warn the user of overwrites.

`exportBIP329` which will take all the entires of transactions,my/externalAddressBook and write them into a List<BIP329>

You can also make the UI components as part of WalletSettings. It doesnt need to have a separate page for now, just two clickable text which opens the file manager to read/write to a fixed expected path. 

It can have an additional text field to show if a bip329 export exists at the expected path and display that to the user.

**Comment by i5hi** [comment]
Gents...We have arrived!
